### PR TITLE
[RLlib] Rename some items in episode APIs.

### PR DESCRIPTION
**Pull request by sven1977**
Rename some items in episode APIs.

* `MultiAgentEpisode._agent_buffered_[actions/rewards/extra_model_outputs]` ->  `_hanging_[actions/rewards/extra_model_outputs]_end`
* Add: `_hanging_[actions/rewards/extra_model_outputs]_begin` to allow for more accurate cuts and slices of MultiAgentEpisode. Right now, when slicing through a sequence of e.g. agent1 NOT receiving observations, but already having acted/received rewards, these hanging actions/rewards should be included in the successor chunk (but at the beginngin). Only this way, proper concatenation can be guaranteed. Otherwise, individual timesteps might get lost entirely.  Note that these properties are not actually used yet in this PR. A follow up PR will fix the behavior of MultiAgentEpisode's `cut()`, `slice()`, and getters.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/multi_agent_episode.py [hunk 1]
```diff
@@ -221,14 +221,21 @@ def __init__(
...
+        # Create caches for hanging actions/rewards/extra_model_outputs.
+        # When an agent gets an observation (and then sends an action), but does not
+        # receive immediately a next observation, we store the "hanging" action (and
+        # related rewards and extra model outputs) in the caches postfixed w/ `_end`
+        # until the next observation is received.
+        self._hanging_actions_end = {}
+        self._hanging_extra_model_outputs_end = defaultdict(dict)
+        self._hanging_rewards_end = defaultdict(float)
+
+        # In case of a `cut()` or `slice()`, we also need to store the hanging actions,
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:233]: Dumb question: Why do we need these extra buffers? Can't we just copy over the `_hanging_actions_end` to the successor? 
When we concatenate them again `self` has these hanging actions and the first observation in the successor episode for the agent will be used to store the now valid transition.

File: rllib/env/multi_agent_episode.py [hunk 2]
```diff
@@ -796,10 +803,10 @@ def cut(self, len_lookback_buffer: int = 0) -> "MultiAgentEpisode":
...
             )
 
-        # If there is data (e.g. actions) in the agents' buffers, we might have to
-        # re-adjust the lookback len further into the past to make sure that these
+        # If there is hanging data (e.g. actions) in the agents' caches, we might have
+        # to re-adjust the lookback len further into the past to make sure that these
         # agents have at least one observation to look back to.
-        for agent_id, agent_actions in self._agent_buffered_actions.items():
+        for agent_id, agent_actions in self._hanging_actions_end.items():
             assert self.env_t_to_agent_t[agent_id].get(-1) == self.SKIP_ENV_TS_TAG
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:810]: Is this always true? What if an agent stepped in the last timestep but did not receive an observation? Do we then denote this as no-step (`SKIP_ENV_TS_TAG`) and count the step at the timestep the missing observation comes in?

File: rllib/env/multi_agent_episode.py [hunk 3]
```diff
@@ -1547,7 +1557,7 @@ def get_state(self) -> Dict[str, Any]:
         Returns: A dicitonary containing pickable data fro a
             `MultiAgentEpisode`.
         """
-        # TODO (simon): Add the buffers.
+        # TODO (simon): Add the agent caches.
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:1560]: Shall we add the caches to the state?

File: rllib/env/multi_agent_episode.py [hunk 4]
```diff
@@ -1621,27 +1631,30 @@ def get_sample_batch(self) -> MultiAgentBatch:
...
         self,
-        consider_buffer: bool = False,
+        consider_hanging_rewards: bool = False,
     ) -> float:
         """Returns all-agent return.
 
         Args:
-            consider_buffer: Whether we should also consider
-                buffered rewards wehn calculating the overall return. Agents might
+            consider_hanging_rewards: Whether we should also consider
```
Comment (simonsays1980) [inline rllib/env/multi_agent_episode.py:1639]: `include_hanging_rewards`?

**Review by simonsays1980 — APPROVED** [review]
LGTM. The set of cached "begin" values appear not to be used.

File: rllib/env/multi_agent_episode.py (hunk 2 above)
Comment (sven1977) [reply rllib/env/multi_agent_episode.py:810]: correct.
As long as the agent does NOT receive a new obs, all steps get marked with "SKIP_ENV_TS_TAG".

File: rllib/env/multi_agent_episode.py (hunk 1 above)
Comment (sven1977) [reply rllib/env/multi_agent_episode.py:233]: Not a dumb question at all!
An example of where it is crucial to have both these chaches:
```
obs = [
   a0: 0, a1: 0   # ts=0
   a0: 1     # ts=1
   a0: 2   # ts=2
]
```

a1's action at ts=0 now is hanging, b/c there is no follow-up observation after its first one. Thus, when we `cut()` (or slice), we put this hanging action in the "before" buffer:

```
successor = MAEps(
   a0: 2,  <- first observation

   self._hanging_actions_before[a1] = [a1's hanging action]
)
```

Then we continue adding timesteps to the succesor chunk (the one returned by `cut()`). Let's assume that in the successor, eventually, a new obs comes in for a1.
```
successor = MAEps(
   a0: 2,  <- first observation
   a0: 3,
   ...
   a0: 5, a1: 5

   self._hanging_actions_before[a1] = [a1's hanging action]
)
```

At this point, we can't do anything with the hanging action (before) as belongs to an observation (the original one of a1 at ts=0 in the previous chunk) that is NOT part of the successor anymore. It's long gone, so to speak. We thus simply leave this hanging action (before) as-is and ONLY use it to check against a preceeding chunk when we do `concat_episodes()`.

Furthermore, assume that we continue the episode and there is another sequence of missing observations for a1, in which case we also now need to store the hanging action at the end for the time when the next obs for a1 DOES come in. However this hanging action at the end should not be confused with the hanging action at the beginning, they are from different timesteps and have nothing to do with each other.

File: rllib/env/multi_agent_episode.py (hunk 4 above)
Comment (sven1977) [reply rllib/env/multi_agent_episode.py:1639]: Fixed in next PR.

**Pull request merged**